File: src/main/java/com/projecttemplategroup/projecttemplateartifact/security/authenticators/SignInAuthenticator.java
package com.projecttemplategroup.projecttemplateartifact.security.authenticators;

import com.projecttemplategroup.projecttemplateartifact.LoggedInDemoUser;
import com.projecttemplategroup.projecttemplateartifact.exceptions.UnauthorizedException;
import com.projecttemplategroup.projecttemplateartifact.security.Pac4jConfig;
import org.hibernate.validator.internal.constraintvalidators.bv.EmailValidator;
import org.pac4j.core.context.WebContext;
import org.pac4j.core.credentials.UsernamePasswordCredentials;
import org.pac4j.core.credentials.authenticator.Authenticator;

public class SignInAuthenticator implements Authenticator<UsernamePasswordCredentials> {

    private void throwsException(String message){
        throw new UnauthorizedException(message);
    }

    @Override
    public void validate(UsernamePasswordCredentials credentials, WebContext context) {
        if (credentials == null) throwsException("No credential");

        String username = credentials.getUsername();
        String password = credentials.getPassword();

        if (new EmailValidator().isValid(username, null)) {
            if(password.equals("qwerty")){

                // todo, its demo sign in logic, use your own implementation with DB
                LoggedInDemoUser user = new LoggedInDemoUser(username);

                credentials.setUserProfile(Pac4jConfig.getProfile(user));
            }else{
                throwsException("Password doesn't much to password qwerty");
            }
        }else{
            throwsException("Username is invalid email");
        }
    }

}

File: src/main/scala/com/projecttemplategroup/projecttemplateartifact/security/authenticators/SignInAuthenticator.scala
package com.projecttemplategroup.projecttemplateartifact.security.authenticators

import com.projecttemplategroup.projecttemplateartifact.exceptions.UnauthorizedException
import com.projecttemplategroup.projecttemplateartifact.security.{Pac4jConfig}
import com.projecttemplategroup.projecttemplateartifact.utils.SecurityUtils
import org.pac4j.core.context.WebContext
import org.pac4j.core.credentials.UsernamePasswordCredentials
import org.pac4j.core.credentials.authenticator.Authenticator

class SignInAuthenticator(var securityUtils: SecurityUtils) extends Authenticator[UsernamePasswordCredentials] {
  private def throwsException(message: String): Unit = {
    throw new UnauthorizedException(message)
  }

  override def validate(credentials: UsernamePasswordCredentials, context: WebContext): Unit = {
    if (credentials == null) throwsException("No credential")
    val username = credentials.getUsername
    val password = credentials.getPassword

    securityUtils.getUserByCredentials(username, password) match {
      case Some(user) => credentials.setUserProfile(Pac4jConfig.loadProfile(user))
      case _ => throwsException("Sign in failed")
    }
  }
}

